src/main/java/fi/aalto/cs/apluscourses/model/MainViewModelUpdater.java

```diff
@@ -0,0 +1,172 @@
+package fi.aalto.cs.apluscourses.model;
+
+import com.intellij.openapi.application.ReadAction;
+import com.intellij.openapi.project.Project;
+
+import fi.aalto.cs.apluscourses.intellij.model.APlusProject;
+import fi.aalto.cs.apluscourses.intellij.model.IntelliJModelFactory;
+import fi.aalto.cs.apluscourses.intellij.services.MainViewModelProvider;
+import fi.aalto.cs.apluscourses.intellij.services.PluginSettings;
+import fi.aalto.cs.apluscourses.presentation.CourseViewModel;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class MainViewModelUpdater {
+
+  @NotNull
+  private MainViewModelProvider mainViewModelProvider;
+
+  @NotNull
+  private APlusProject aplusProject;
+
+  private long updateInterval;
+
+  /**
+   * Construct a {@link MainViewModelUpdater} with the given {@link MainViewModelProvider}, project,
+   * and update interval.
+   * @param mainViewModelProvider The {@link MainViewModelProvider} from which the main view model
+   *                              to be updated is gotten.
+   * @param project               The project to which this updater is tied.
+   * @param updateInterval        The interval at which the updater performs the update.
+   */
+  public MainViewModelUpdater(@NotNull MainViewModelProvider mainViewModelProvider,
+                              @NotNull Project project,
+                              @NotNull long updateInterval) {
+    this.mainViewModelProvider = mainViewModelProvider;
+    this.aplusProject = new APlusProject(project);
+    this.updateInterval = updateInterval;
+  }
+
+  /**
+   * Construct a {@link MainViewModelUpdater} with the given project.
+   * @param project The project to which thsi updater is tied.
+   */
+  public MainViewModelUpdater(@NotNull Project project) {
+    this(PluginSettings.getInstance(), project, PluginSettings.MAIN_VIEW_MODEL_UPDATE_INTERVAL);
+  }
+
+  private void interruptIfProjectClosed() throws InterruptedException {
+    if (aplusProject.getProject().isDisposed() || !aplusProject.getProject().isOpen()) {
+      throw new InterruptedException();
+    }
+  }
+
+  @NotNull
+  URL getCourseUrl() throws InterruptedException {
+    return ReadAction.compute(() -> {
+      interruptIfProjectClosed();
+      try {
+        URL url = aplusProject.getCourseFileUrl();
+        if (url == null) {
+          throw new InterruptedException();
```

Comment:
Is this correct way to handle this? Where is this caught? How do we recover from this? (The same question applies other places where this exception is thrown.)

Reply:
The exception here is perhaps confusing, but there is nothing special about `InterruptedException`. There is a single catch clause for `InterruptedException`, which is outside the infinite loop, so the thread terminates after that. The catch clause also calls `Thread.currentThread().interrupt()`, which makes it "visible" that this thread was interrupted (currently that information isn't used anywhere).

`Thread.sleep` throws an `InterruptedException`, so I decided to have all the possible interruptions (e.g. project was closed) throw an `InterruptedException` and handle it in a single catch clause.

https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html